lib/Dialect/FIRRTL/Transforms/AddSeqMemPorts.cpp

```diff
@@ -403,27 +417,75 @@ void AddSeqMemPortsPass::runOnOperation() {
   }
   extraPortsAttr = ArrayAttr::get(context, extraPorts);
 
+  auto *effectiveDut = instanceInfo->getEffectiveDut();
+
   // If there are no user ports, don't do anything.
   if (!userPorts.empty()) {
     // Update ports statistic.
     numAddedPorts += userPorts.size();
 
-    // Visit the nodes in post-order.
-    for (auto *node : llvm::post_order(dutNode)) {
+    // Visit the modules in post-order in one of two ways:
+    //
+    //   1. If no design-under-test is specified, visit all modules.
+    //   2. If the design-under-test is specified, only visit modules under it.
+    //
+    // Additionally, skip any modules that are wholly instantiated under layers.
+    // If any memories are partially instantiated under a layer then error.
+    for (auto *node : llvm::post_order(instanceGraph)) {
       auto op = node->getModule();
+      // Skip the module if there is a DUT and no instances are under it.
+      if (instanceInfo->hasDut() && !instanceInfo->anyInstanceUnderDut(op))
+        continue;
+
+      // Skip anything wholly under a layer.
+      if (instanceInfo->allInstancesUnderLayer(op))
+        continue;
```

Comment:
If the DUT is wholly under a layer, then we won't add any MBIST ports. Seems OK.

Reply:
That seems reasonable, though I suppose it could work? E.g., there's no problem wiring to the DUT as long as you don't have to pass through any intermediary layers to do it. (InstanceInfo doesn't track that, though!)

This smells like a missing check that should be in `LowerAnnotations` (as I don't think we can do it with a verifier).